pkg/controllers/monitoring-stack/controller.go

```diff
@@ -164,10 +169,65 @@ func (r *reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 
 		}
 	}
+	return r.setupFinalizer(ctx, ms)
+}
+
+func (r *reconciler) deleteGrafanaDS(ctx context.Context, ms *stack.MonitoringStack) error {
+	logger := r.logger.WithValues(stackName(ms)...)
+
+	gds := types.NamespacedName{Namespace: goctrl.Namespace, Name: GrafanaDSName(ms)}
+	grafanaDS := grafanav1alpha1.GrafanaDataSource{}
+	if err := r.k8sClient.Get(ctx, gds, &grafanaDS); err != nil {
+		// if the datasource is already deleted, take no further action
+		return client.IgnoreNotFound(err)
+	}
 
+	// grafana ds exists; so delete it
+	logger.WithValues("GrafanaDataSource", grafanaDS.Name).Info("Deleting GrafanaDataSource")
+	err := r.k8sClient.Delete(ctx, &grafanaDS)
+	return client.IgnoreNotFound(err)
+}
+
+func (r *reconciler) cleanupResources(ctx context.Context, ms *stack.MonitoringStack) (ctrl.Result, error) {
+	logger := r.logger.WithValues(stackName(ms)...)
+
+	if !controllerutil.ContainsFinalizer(ms, finalizerName) {
+		logger.V(6).Info("Finalizer already removed")
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.deleteGrafanaDS(ctx, ms); err != nil {
+		logger.V(6).Info("Could not delete GrafanaDataSource", "err", err)
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("Removing finalizer")
+	controllerutil.RemoveFinalizer(ms, finalizerName)
+	err := r.k8sClient.Update(ctx, ms)
+	if errors.IsConflict(err) {
+		return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+	}
+	return ctrl.Result{}, err
+}
+
+func (r *reconciler) setupFinalizer(ctx context.Context, ms *stack.MonitoringStack) (ctrl.Result, error) {
```

Comment:
@prashbnair I think it is worth having  a test that validates that the finalizer is set on a monitoring stack 

Note ... This can be done is a follow up PR as well.